aixplain/api/base.py

```diff
@@ -0,0 +1,317 @@
+        else:
+            raise ValueError(f"Invalid param type: {self.param_type}")
+
+    def link(self, to_param: "Param") -> "Param":
+        """
+        Link the output of the param to the input of another param.
+        :param to_param: the input param
+        :return: the param
+        """
+        assert to_param.param_type == ParamType.INPUT, "Invalid param type"
+        assert (
+            self.node and self in self.node.outputValues
+        ), "Param not attached to a node"
+        to_param.back_link(self)
+
+    def back_link(self, from_param: "Param") -> "Param":
+        """
+        Link the input of the param to the output of another param.
+        :param from_param: the output param
+        :return: the param
+        """
+        assert from_param.param_type == ParamType.OUTPUT, "Invalid param type"
+        assert (
+            self.node and self in self.node.inputValues
+        ), "Param not attached to a node"
+        from_param.node.link(self.node, from_param.code, self.code)
+
+
+@dataclass
+class InputParam(Param):
+
+    param_type: ParamType = ParamType.INPUT
+    is_required: InitVar[bool] = True
+
+    def __post_init__(self, node: "Node" = None, is_required: bool = False):
+        """
+        Post init method to set the required flag.
+        """
+        super().__post_init__(node=node)
+        self.is_required = is_required
+
+
+@dataclass
+class OutputParam(Param):
+
+    param_type: ParamType = ParamType.OUTPUT
+
+
+@dataclass
+class ParamMapping:
+    """
+    Param mapping class, this class will be used to map the output of the
+    node to the input of another node.
+    """
+
+    from_param: Union[str, Param]
+    to_param: Union[str, Param]
+
+    def __post_init__(self):
+        """
+        Post init method to convert the params to param codes if they are
+        params.
+        """
+        if isinstance(self.from_param, Param):
+            self.from_param = self.from_param.code
+        if isinstance(self.to_param, Param):
+            self.to_param = self.to_param.code
+
+
+@dataclass
+class Link:
+    """
+    Link class, this class will be used to link the output of the node to the
+    input of another node.
+    """
+
+    from_node: Union[int, "Node"]
+    to_node: Union[int, "Node"]
+    paramMapping: List[ParamMapping] = field(default_factory=list)
+    pipeline: InitVar["Pipeline"] = None
+
+    def __post_init__(self, pipeline: "Pipeline" = None):
+        if pipeline:
+            self.attach(pipeline)
+        if isinstance(self.from_node, Node):
+            self.from_node = self.from_node.number
+        if isinstance(self.to_node, Node):
+            self.to_node = self.to_node.number
+
+    def attach(self, pipeline: "Pipeline"):
+        """
+        Attach the link to the pipeline.
+        :param pipeline: the pipeline
+        """
+        assert not self.pipeline, "Link already attached to a pipeline"
+
+        self.pipeline = pipeline
```

Comment:
Why do we need to keep an instance of pipeline inside every node and link?

Reply:
Because any node can be instantiated with a pipeline instance so that it can add itself to pipeline nodes:

All below the methods are valid for relating any node with a pipeline.
``` python
pipeline = Pipeline()
script = Script(script_path='myscript.py')
script.attach(pipeline)
```
or 
``` python
pipeline = Pipeline()
script = Script(script_path='myscript.py', pipeline=pipeline)
```
or
``` python
pipeline = Pipeline()
script = Script(script_path='myscript.py')
pipeline.add_node(script)
```